File: nodejs/gajihub.mjs
/**
 * ============================================================================
 *  gajihub.mjs — Fungsi bantu yang dipakai oleh semua demo Node.js
 * ============================================================================
 *
 *  File ini di-"import" oleh setiap file demo. Anda TIDAK perlu mengubahnya.
 *  Alamat API dan token diatur di file .env di folder utama gajihub-api-demo/
 *  (salin dari .env.example). File .env yang sama dipakai demo PHP & Python.
 *
 *  Tanpa npm install: hanya memakai fitur bawaan Node.js 18 atau lebih baru.
 *
 *  Daftar fungsi (semua fungsi api* dan unduhFile dipanggil dengan await):
 *    mulaiDemo('Judul')                  -> tampilkan judul demo
 *    apiGet('/endpoint', {parameter})    -> ambil data            (GET)
 *    apiPost('/endpoint', {data})        -> tambah data           (POST)
 *    apiPut('/endpoint', {data})         -> ubah data             (PUT)
 *    apiPatch('/endpoint', {data})       -> ubah sebagian data    (PATCH)
 *    apiDelete('/endpoint')              -> hapus data            (DELETE)
 *    unduhFile('/endpoint', {parameter}) -> unduh file Excel/CSV ke folder hasil-unduhan/
 *    tampilkan(hasil)                    -> cetak hasil request ke layar
 *    tulis('teks')                       -> cetak satu baris teks
 *
 *  Fungsi tanggal (zona waktu Asia/Jakarta):
 *    tanggal()     -> hari ini, YYYY-MM-DD   tanggal(-1) = kemarin, tanggal(7) = 7 hari lagi
 *    awalBulan()   -> YYYY-MM-01             akhirBulan() -> tanggal terakhir bulan ini
 *    bulanIni()    -> YYYY-MM                bulanLalu()  -> YYYY-MM bulan lalu
 *
 *  Setiap fungsi api* mengembalikan object:
 *    {
 *      status: 200,      // kode HTTP (200 = berhasil)
 *      sukses: true,     // true jika status 2xx
 *      data:   {...},    // isi respons JSON yang sudah diubah jadi object JavaScript
 *      isi:    Buffer,   // isi respons mentah (isi file)
 *      header: Headers,  // header respons
 *    }
 * ============================================================================
 */

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

// Folder utama gajihub-api-demo/ (satu tingkat di atas folder nodejs/)
const FOLDER_UTAMA = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');

// Folder tempat file hasil export disimpan
const FOLDER_UNDUHAN = path.join(FOLDER_UTAMA, 'hasil-unduhan');

// ----------------------------------------------------------------------------
// 1. Konfigurasi dari file .env
// ----------------------------------------------------------------------------

let isiEnv = null;

/**
 * Ambil nilai konfigurasi dari file .env, misalnya konfigurasi('API_HOST').
 */
function konfigurasi(nama) {
    if (isiEnv === null) {
        const fileEnv = path.join(FOLDER_UTAMA, '.env');

        if (!fs.existsSync(fileEnv)) {
            berhenti('File .env belum ada. Salin .env.example menjadi .env, lalu isi API_HOST dan ACCESS_TOKEN.');
        }

        isiEnv = {};
        for (let baris of fs.readFileSync(fileEnv, 'utf8').split(/\r?\n/)) {
            baris = baris.trim();

            // Lewati komentar (#) dan baris tanpa tanda =
            if (baris === '' || baris.startsWith('#') || !baris.includes('=')) {
                continue;
            }

            const posisi = baris.indexOf('=');
            const kunci = baris.slice(0, posisi).trim();
            const nilai = baris.slice(posisi + 1).trim().replace(/^["']+|["']+$/g, '');
            isiEnv[kunci] = nilai;
        }
    }

    if (!isiEnv[nama]) {
        berhenti(`${nama} belum diisi di file .env`);
    }

    return isiEnv[nama];
}

// ----------------------------------------------------------------------------
// 2. Mengirim request ke API GajiHub
// ----------------------------------------------------------------------------

export const apiGet = (endpoint, parameter = {}) => kirimRequest('GET', endpoint, parameter);
export const apiPost = (endpoint, data = {}) => kirimRequest('POST', endpoint, data);
export const apiPut = (endpoint, data = {}) => kirimRequest('PUT', endpoint, data);
export const apiPatch = (endpoint, data = {}) => kirimRequest('PATCH', endpoint, data);
export const apiDelete = (endpoint, parameter = {}) => kirimRequest('DELETE', endpoint, parameter);

/**
 * Ubah object parameter menjadi teks query URL, contoh: page=1&ids[]=1&ids[]=2
 */
function buatQuery(parameter) {
    const query = new URLSearchParams();
    for (const [kunci, nilai] of Object.entries(parameter)) {
        if (nilai === null || nilai === undefined) {
            continue; // nilai kosong tidak dikirim
        }
        if (Array.isArray(nilai)) {
            nilai.forEach((isi) => query.append(`${kunci}[]`, isi));
        } else if (typeof nilai === 'boolean') {
            query.append(kunci, nilai ? 1 : 0);
        } else {
            query.append(kunci, nilai);
        }
    }
    return query.toString();
}

/**
 * Inti dari semua request. Semua endpoint GajiHub diawali /hr,
 * jadi endpoint cukup ditulis mulai dari /hr/...
 */
async function kirimRequest(method, endpoint, data = {}) {
    let url = konfigurasi('API_HOST').replace(/\/+$/, '') + endpoint;

    const header = {
        Accept: 'application/json',
        Authorization: 'Bearer ' + konfigurasi('ACCESS_TOKEN'),
    };

    // GET/DELETE: parameter dikirim lewat URL (?page=1&per_page=10)
    // POST/PUT/PATCH: data dikirim sebagai JSON di body request
    const kirimSebagaiJson = ['POST', 'PUT', 'PATCH'].includes(method);
    let body;

    if (kirimSebagaiJson) {
        header['Content-Type'] = 'application/json';
        body = JSON.stringify(data);
    } else if (Object.keys(data).length > 0) {
        url += '?' + buatQuery(data);
    }

    tulis(`${method} ${url}`);

    let respons;
    try {
        respons = await fetch(url, {
            method,
            headers: header,
            body,
            signal: AbortSignal.timeout(300_000), // export bisa lama, tunggu maksimal 5 menit
        });
    } catch (gagal) {
        berhenti('Gagal terhubung ke server: ' + (gagal.cause?.message ?? gagal.message));
    }

    const isi = Buffer.from(await respons.arrayBuffer());

    let dataJson = null;
    try {
        dataJson = JSON.parse(isi.toString('utf8'));
    } catch {
        // bukan JSON, misalnya isi file Excel
    }

    return {
        status: respons.status,
        sukses: respons.status >= 200 && respons.status < 300,
        data: dataJson,
        isi,
        header: respons.headers,
    };
}

// ----------------------------------------------------------------------------
// 3. Mengunduh file export (Excel / CSV)
// ----------------------------------------------------------------------------

/**
 * Unduh file export lalu simpan ke folder hasil-unduhan/.
 * Mengembalikan lokasi file, atau null jika gagal / tidak ada data.
 */
export async function unduhFile(endpoint, parameter = {}) {
    const hasil = await apiGet(endpoint, parameter);

    // Jika server membalas JSON, berarti ada pesan error (misalnya format tanggal salah)
    if (!hasil.sukses || hasil.data !== null) {
        tampilkan(hasil);
        return null;
    }

    // Respons kosong = tidak ada data pada periode/filter tersebut
    if (hasil.isi.length === 0) {
        tulis('Tidak ada data untuk diexport pada periode/filter tersebut.');
        return null;
    }

    // Ambil nama file dari header "Content-Disposition: attachment; filename=..."
    let namaFile = 'export_' + tanggal().replaceAll('-', '') + '_' + Date.now();
    const cocok = /filename="?([^";]+)"?/i.exec(hasil.header.get('content-disposition') ?? '');
    if (cocok) {
        namaFile = path.basename(cocok[1]);
    }

    fs.mkdirSync(FOLDER_UNDUHAN, { recursive: true });

    const lokasiFile = path.join(FOLDER_UNDUHAN, namaFile);
    fs.writeFileSync(lokasiFile, hasil.isi);

    tulis(`Berhasil! File disimpan di: ${lokasiFile} (${hasil.isi.length} bytes)`);

    return lokasiFile;
}

// ----------------------------------------------------------------------------
// 4. Menampilkan hasil
// ----------------------------------------------------------------------------

/**
 * Tampilkan judul demo.
 */
export function mulaiDemo(judul) {
    tulis('');
    tulis(`=== ${judul} ===`);
}

/**
 * Cetak hasil request: status HTTP + isi respons dalam format JSON yang rapi.
 */
export function tampilkan(hasil) {
    tulis(`Status HTTP: ${hasil.status}` + (hasil.sukses ? ' (berhasil)' : ' (gagal)'));

    if (!hasil.sukses) {
        tulis(artiStatus(hasil.status));
    }

    tulis(hasil.data !== null ? JSON.stringify(hasil.data, null, 4) : hasil.isi.toString('utf8'));
}

/**
 * Cetak satu baris teks.
 */
export function tulis(teks) {
    console.log(teks);
}

/**
 * Penjelasan singkat untuk kode status HTTP yang sering muncul.
 */
function artiStatus(status) {
    const daftar = {
        400: 'Artinya: parameter/data yang dikirim tidak valid. Lihat pesan di bawah.',
        401: 'Artinya: token tidak valid atau sudah kedaluwarsa. Periksa ACCESS_TOKEN di file .env.',
        403: 'Artinya: user pemilik token tidak punya hak akses ke fitur ini.',
        404: 'Artinya: data atau endpoint tidak ditemukan. Periksa ID / alamat endpoint.',
        429: 'Artinya: terlalu banyak request. Tunggu sebentar lalu coba lagi.',
    };
    return daftar[status] ?? (status >= 500 ? 'Artinya: terjadi kesalahan di server. Coba lagi beberapa saat.' : '');
}

/**
 * Tampilkan pesan lalu hentikan program.
 */
function berhenti(pesan) {
    tulis('ERROR: ' + pesan);
    process.exit(1);
}

// ----------------------------------------------------------------------------
// 5. Fungsi tanggal (zona waktu Asia/Jakarta)
// ----------------------------------------------------------------------------

const duaDigit = (angka) => String(angka).padStart(2, '0');

/** Waktu sekarang di Asia/Jakarta (UTC+7), dibaca dengan getUTC...() */
const waktuJakarta = (tambahHari = 0) => new Date(Date.now() + (7 * 3600 + tambahHari * 86400) * 1000);

/** Tanggal hari ini (YYYY-MM-DD). tanggal(-1) = kemarin, tanggal(7) = 7 hari lagi. */
export function tanggal(tambahHari = 0) {
    const waktu = waktuJakarta(tambahHari);
    return `${waktu.getUTCFullYear()}-${duaDigit(waktu.getUTCMonth() + 1)}-${duaDigit(waktu.getUTCDate())}`;
}

/** Bulan ini (YYYY-MM). */
export const bulanIni = () => tanggal().slice(0, 7);

/** Tanggal pertama bulan ini (YYYY-MM-01). */
export const awalBulan = () => bulanIni() + '-01';

/** Tanggal terakhir bulan ini (YYYY-MM-DD). */
export function akhirBulan() {
    const waktu = waktuJakarta();
    const hariTerakhir = new Date(Date.UTC(waktu.getUTCFullYear(), waktu.getUTCMonth() + 1, 0)).getUTCDate();
    return `${bulanIni()}-${duaDigit(hariTerakhir)}`;
}

/** Bulan lalu (YYYY-MM). */
export function bulanLalu() {
    const waktu = waktuJakarta();
    const awalBulanLalu = new Date(Date.UTC(waktu.getUTCFullYear(), waktu.getUTCMonth() - 1, 1));
    return `${awalBulanLalu.getUTCFullYear()}-${duaDigit(awalBulanLalu.getUTCMonth() + 1)}`;
}

File: python/gajihub.py
"""
============================================================================
 gajihub.py — Fungsi bantu yang dipakai oleh semua demo Python
============================================================================

 File ini di-"import" oleh setiap file demo. Anda TIDAK perlu mengubahnya.
 Alamat API dan token diatur di file .env di folder utama gajihub-api-demo/
 (salin dari .env.example). File .env yang sama dipakai demo PHP & Node.js.

 Hanya memakai library bawaan Python (tanpa pip install).

 Daftar fungsi:
   mulai_demo('Judul')                  -> tampilkan judul demo
   api_get('/endpoint', {parameter})    -> ambil data            (GET)
   api_post('/endpoint', {data})        -> tambah data           (POST)
   api_put('/endpoint', {data})         -> ubah data             (PUT)
   api_patch('/endpoint', {data})       -> ubah sebagian data    (PATCH)
   api_delete('/endpoint')              -> hapus data            (DELETE)
   unduh_file('/endpoint', {parameter}) -> unduh file Excel/CSV ke folder hasil-unduhan/
   tampilkan(hasil)                     -> cetak hasil request ke layar
   tulis('teks')                        -> cetak satu baris teks

 Fungsi tanggal (zona waktu Asia/Jakarta):
   tanggal()      -> hari ini, YYYY-MM-DD   tanggal(-1) = kemarin, tanggal(7) = 7 hari lagi
   awal_bulan()   -> YYYY-MM-01             akhir_bulan() -> tanggal terakhir bulan ini
   bulan_ini()    -> YYYY-MM                bulan_lalu()  -> YYYY-MM bulan lalu

 Setiap fungsi api_* mengembalikan dict:
   {
     'status': 200,       # kode HTTP (200 = berhasil)
     'sukses': True,      # True jika status 2xx
     'data':   {...},     # isi respons JSON yang sudah diubah jadi dict Python
     'isi':    b'...',    # isi respons mentah (bytes / isi file)
     'header': {...},     # header respons (nama header huruf kecil)
   }
============================================================================
"""

import calendar
import json
import os
import re
import sys
import urllib.error
import urllib.parse
import urllib.request
from datetime import datetime, timedelta, timezone
from pathlib import Path

# Folder utama gajihub-api-demo/ (satu tingkat di atas folder python/)
FOLDER_UTAMA = Path(__file__).resolve().parent.parent

# Folder tempat file hasil export disimpan
FOLDER_UNDUHAN = FOLDER_UTAMA / "hasil-unduhan"

# Zona waktu Asia/Jakarta (WIB, UTC+7)
WIB = timezone(timedelta(hours=7))

# Agar huruf non-ASCII tampil benar di terminal Windows
if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")

# ----------------------------------------------------------------------------
# 1. Konfigurasi dari file .env
# ----------------------------------------------------------------------------

_isi_env = None


def konfigurasi(nama):
    """Ambil nilai konfigurasi dari file .env, misalnya konfigurasi('API_HOST')."""
    global _isi_env

    if _isi_env is None:
        file_env = FOLDER_UTAMA / ".env"

        if not file_env.is_file():
            berhenti("File .env belum ada. Salin .env.example menjadi .env, lalu isi API_HOST dan ACCESS_TOKEN.")

        _isi_env = {}
        for baris in file_env.read_text(encoding="utf-8").splitlines():
            baris = baris.strip()

            # Lewati komentar (#) dan baris tanpa tanda =
            if baris == "" or baris.startswith("#") or "=" not in baris:
                continue

            kunci, nilai = baris.split("=", 1)
            _isi_env[kunci.strip()] = nilai.strip().strip("\"'")

    if not _isi_env.get(nama):
        berhenti(f"{nama} belum diisi di file .env")

    return _isi_env[nama]


# ----------------------------------------------------------------------------
# 2. Mengirim request ke API GajiHub
# ----------------------------------------------------------------------------

def api_get(endpoint, parameter=None):
    return kirim_request("GET", endpoint, parameter)


def api_post(endpoint, data=None):
    return kirim_request("POST", endpoint, data)


def api_put(endpoint, data=None):
    return kirim_request("PUT", endpoint, data)


def api_patch(endpoint, data=None):
    return kirim_request("PATCH", endpoint, data)


def api_delete(endpoint, parameter=None):
    return kirim_request("DELETE", endpoint, parameter)


def buat_query(parameter):
    """Ubah dict parameter menjadi teks query URL, contoh: page=1&ids[]=1&ids[]=2"""
    pasangan = []
    for kunci, nilai in parameter.items():
        if nilai is None:
            continue  # nilai kosong tidak dikirim
        if isinstance(nilai, (list, tuple)):
            pasangan += [(f"{kunci}[]", isi) for isi in nilai]
        elif isinstance(nilai, bool):
            pasangan.append((kunci, int(nilai)))
        else:
            pasangan.append((kunci, nilai))
    return urllib.parse.urlencode(pasangan)


def kirim_request(method, endpoint, data=None):
    """
    Inti dari semua request. Semua endpoint GajiHub diawali /hr,
    jadi endpoint cukup ditulis mulai dari /hr/...
    """
    data = data or {}
    url = konfigurasi("API_HOST").rstrip("/") + endpoint

    header = {
        "Accept": "application/json",
        "Authorization": "Bearer " + konfigurasi("ACCESS_TOKEN"),
    }

    # GET/DELETE: parameter dikirim lewat URL (?page=1&per_page=10)
    # POST/PUT/PATCH: data dikirim sebagai JSON di body request
    body = None
    if method in ("POST", "PUT", "PATCH"):
        header["Content-Type"] = "application/json"
        body = json.dumps(data).encode("utf-8")
    elif data:
        url += "?" + buat_query(data)

    tulis(f"{method} {url}")

    request = urllib.request.Request(url, data=body, headers=header, method=method)

    try:
        # Export bisa lama, tunggu maksimal 5 menit
        with urllib.request.urlopen(request, timeout=300) as respons:
            status, isi, header_respons = respons.status, respons.read(), respons.headers
    except urllib.error.HTTPError as gagal:
        # Status 4xx / 5xx tetap dibaca isinya (berisi pesan error dari server)
        status, isi, header_respons = gagal.code, gagal.read(), gagal.headers
    except (urllib.error.URLError, OSError) as gagal:
        berhenti(f"Gagal terhubung ke server: {gagal}")

    try:
        data_json = json.loads(isi)
    except ValueError:
        data_json = None  # bukan JSON, misalnya isi file Excel

    return {
        "status": status,
        "sukses": 200 <= status < 300,
        "data": data_json,
        "isi": isi,
        "header": {kunci.lower(): nilai for kunci, nilai in header_respons.items()},
    }


# ----------------------------------------------------------------------------
# 3. Mengunduh file export (Excel / CSV)
# ----------------------------------------------------------------------------

def unduh_file(endpoint, parameter=None):
    """
    Unduh file export lalu simpan ke folder hasil-unduhan/.
    Mengembalikan lokasi file, atau None jika gagal / tidak ada data.
    """
    hasil = api_get(endpoint, parameter)

    # Jika server membalas JSON, berarti ada pesan error (misalnya format tanggal salah)
    if not hasil["sukses"] or hasil["data"] is not None:
        tampilkan(hasil)
        return None

    # Respons kosong = tidak ada data pada periode/filter tersebut
    if hasil["isi"] == b"":
        tulis("Tidak ada data untuk diexport pada periode/filter tersebut.")
        return None

    # Ambil nama file dari header "Content-Disposition: attachment; filename=..."
    nama_file = "export_" + datetime.now(WIB).strftime("%Y%m%d_%H%M%S")
    cocok = re.search(r'filename="?([^";]+)"?', hasil["header"].get("content-disposition", ""), re.IGNORECASE)
    if cocok:
        nama_file = os.path.basename(cocok.group(1))

    FOLDER_UNDUHAN.mkdir(parents=True, exist_ok=True)

    lokasi_file = FOLDER_UNDUHAN / nama_file
    lokasi_file.write_bytes(hasil["isi"])

    tulis(f"Berhasil! File disimpan di: {lokasi_file} ({len(hasil['isi'])} bytes)")

    return lokasi_file


# ----------------------------------------------------------------------------
# 4. Menampilkan hasil
# ----------------------------------------------------------------------------

def mulai_demo(judul):
    """Tampilkan judul demo."""
    tulis("")
    tulis(f"=== {judul} ===")


def tampilkan(hasil):
    """Cetak hasil request: status HTTP + isi respons dalam format JSON yang rapi."""
    tulis(f"Status HTTP: {hasil['status']}" + (" (berhasil)" if hasil["sukses"] else " (gagal)"))

    if not hasil["sukses"]:
        tulis(arti_status(hasil["status"]))

    if hasil["data"] is not None:
        teks = json.dumps(hasil["data"], indent=4, ensure_ascii=False)
    else:
        teks = hasil["isi"].decode("utf-8", errors="replace")

    tulis(teks)


def tulis(teks):
    """Cetak satu baris teks."""
    print(teks)


def arti_status(status):
    """Penjelasan singkat untuk kode status HTTP yang sering muncul."""
    daftar = {
        400: "Artinya: parameter/data yang dikirim tidak valid. Lihat pesan di bawah.",
        401: "Artinya: token tidak valid atau sudah kedaluwarsa. Periksa ACCESS_TOKEN di file .env.",
        403: "Artinya: user pemilik token tidak punya hak akses ke fitur ini.",
        404: "Artinya: data atau endpoint tidak ditemukan. Periksa ID / alamat endpoint.",
        429: "Artinya: terlalu banyak request. Tunggu sebentar lalu coba lagi.",
    }
    if status in daftar:
        return daftar[status]
    return "Artinya: terjadi kesalahan di server. Coba lagi beberapa saat." if status >= 500 else ""


def berhenti(pesan):
    """Tampilkan pesan lalu hentikan program."""
    tulis("ERROR: " + pesan)
    sys.exit(1)


# ----------------------------------------------------------------------------
# 5. Fungsi tanggal (zona waktu Asia/Jakarta)
# ----------------------------------------------------------------------------

def tanggal(tambah_hari=0):
    """Tanggal hari ini (YYYY-MM-DD). tanggal(-1) = kemarin, tanggal(7) = 7 hari lagi."""
    return (datetime.now(WIB) + timedelta(days=tambah_hari)).strftime("%Y-%m-%d")


def bulan_ini():
    """Bulan ini (YYYY-MM)."""
    return datetime.now(WIB).strftime("%Y-%m")


def awal_bulan():
    """Tanggal pertama bulan ini (YYYY-MM-01)."""
    return bulan_ini() + "-01"


def akhir_bulan():
    """Tanggal terakhir bulan ini (YYYY-MM-DD)."""
    sekarang = datetime.now(WIB)
    return f"{bulan_ini()}-{calendar.monthrange(sekarang.year, sekarang.month)[1]:02d}"


def bulan_lalu():
    """Bulan lalu (YYYY-MM)."""
    return (datetime.now(WIB).replace(day=1) - timedelta(days=1)).strftime("%Y-%m")
